### PR TITLE
fix hexadecimal literal value conversion

### PR DESCRIPTION
**Pull request by jennifersp**
MySQL docs:
`In numeric contexts, MySQL treats a hexadecimal literal like a BIGINT UNSIGNED (64-bit unsigned integer).`

### CODE REVIEW

File: enginetest/enginetests.go [hunk 1]
```diff
@@ -4401,6 +4401,25 @@ func TestPreparedInsert(t *testing.T, harness Harness) {
...
+			Name: "simple decimal type insert",
+			SetUpScript: []string{
+				"CREATE TABLE test(id int primary key auto_increment, decimal_test DECIMAL(9,2), decimal_test_2 DECIMAL(9,2), decimal_test_3 DECIMAL(9,2))",
+			},
+			Assertions: []queries.ScriptTestAssertion{
+				{
+					Query: "INSERT INTO test(decimal_test, decimal_test_2, decimal_test_3) VALUES (?, ?, ?)",
+					Bindings: map[string]sql.Expression{
+						"v1": expression.NewLiteral(10, sql.Int64),
+						"v2": expression.NewLiteral([]byte("10.5"), sql.MustCreateString(sqltypes.VarBinary, 4, sql.Collation_binary)),
```
Comment (fulghum) [inline enginetest/enginetests.go:4414]: Does the `varbinary` type trigger the same issue as `X'10'`? Just curious if it's worth adding that hex value to the test.

File: sql/expression/convert.go [hunk 2]
```diff
@@ -240,3 +246,18 @@ func convertValue(val interface{}, castTo string, originType sql.Type) (interfac
...
 }
+
+// convertHexBlobToDecimalForNumericContext returns numeric value if the value is binary and origin type is blob
+// IF the convertTo type is number type. This applies only for hex LITERAL values parsed it is parsed as
+// BLOB type in the parser but to binary string type.
+// -- other byte arrays of other sql types does not apply here.
+func convertHexBlobToDecimalForNumericContext(val interface{}, originType sql.Type) interface{} {
+	if bin, isBinary := val.([]byte); isBinary && sql.IsBlobType(originType) {
+		stringVal := hex.EncodeToString(bin)
+		decimalNum, err := strconv.ParseUint(stringVal, 16, 64)
```
Comment (fulghum) [inline sql/expression/convert.go:257]: Would be good to do something with this `err` so we don't swallow any errors and make future debugging harder. Think this should error the statement if this happens?

**Review by fulghum — COMMENTED** [review]
Nice work digging in and finding this fix! Code looks good, I just had two small questions and wanted to take one more quick look in the morning when my brain is fresh.

File: sql/expression/convert.go [hunk 3]
```diff
@@ -240,3 +246,18 @@ func convertValue(val interface{}, castTo string, originType sql.Type) (interfac
 		return nil, nil
 	}
 }
+
+// convertHexBlobToDecimalForNumericContext returns numeric value if the value is binary and origin type is blob
```
Comment (PavelSafronov) [inline sql/expression/convert.go:250]: It's not clear to me what we are saying here. Can you rephrase/rewrite this?

File: sql/expression/convert.go [hunk 4]
```diff
@@ -240,3 +246,18 @@ func convertValue(val interface{}, castTo string, originType sql.Type) (interfac
...
+// IF the convertTo type is number type. This applies only for hex LITERAL values parsed it is parsed as
+// BLOB type in the parser but to binary string type.
+// -- other byte arrays of other sql types does not apply here.
```
Comment (PavelSafronov) [inline sql/expression/convert.go:253]: Is `--` necessary in the comments here?
And "does not apply here" doesn't accurately convey what happens.

Maybe something like this?
```
// Byte arrays of other SQL types are not handled here.
```

File: sql/type.go [hunk 5]
```diff
@@ -567,7 +567,7 @@ func IsByteType(t Type) bool {
 	}
 }
 
-// IsBinaryType checks if t is BINARY, VARBINARY, or BLOB
+// IsBinaryType checks if it is BINARY, VARBINARY, or BLOB
```
Comment (PavelSafronov) [inline sql/type.go:570]: `t` is the proper text here, we're referring to the variable. Don't think this change is needed.
Same for the change on line 560.

File: sql/decimal.go [hunk 6]
```diff
@@ -241,12 +240,11 @@ func (t decimalType) ConvertToNullDecimal(v interface{}) (decimal.NullDecimal, e
 	case decimal.Decimal:
 		res = value
 	case []uint8:
-		var err error
-		val, err := strconv.ParseUint(hex.EncodeToString(value), 16, 64)
+		val, err := strconv.ParseFloat(string(value[:]), 64)
```
Comment (PavelSafronov) [inline sql/decimal.go:243]: `[:]` is not necessary here, as it just returns `value`.

File: enginetest/enginetests.go (hunk 1 above)
Comment (jennifersp) [reply enginetest/enginetests.go:4414]: Do you mean to test with `X'10'` instead of `10.5` as binding value?

Something I noticed that I did not understand clearly was that we convert many types into `varbinary` in https://github.com/dolthub/vitess/blob/main/go/mysql/query.go#L842 , but we are not able to convert them back to its original type in this method, https://github.com/dolthub/go-mysql-server/blob/main/server/handler.go#L180 , so for example passing in `X'10'` causes panic because we try to handle it as the same way as any string like `"X'10'"`

Comment (fulghum) [reply enginetest/enginetests.go:4414]: Interesting... sending a hex string as a bindvar still causes a panic? Could you please open a tracking issue for that one with a good repro and we can tackle it separately?

**Review by fulghum — APPROVED** [review]
Looks good. Thanks for iterating on the feedback! 

I'm still concerned about that panic with hex formatted string bind vars, but we can tackle that one separately.

File: enginetest/enginetests.go (hunk 1 above)
Comment (jennifersp) [reply enginetest/enginetests.go:4414]: made an issue here https://github.com/dolthub/dolt/issues/4989
